bin/propolis-server/src/lib/vm/request_queue.rs

```diff
@@ -0,0 +1,358 @@
+            // performed.
+            ExternalRequest::Reboot => {
+                assert!(matches!(self.allowed.start, Disposition::Ignore));
+                assert!(!matches!(
+                    self.allowed.migrate_as_target,
+                    Disposition::Enqueue
+                ));
+            }
+
+            // Queueing a request to stop an instance disables any other
+            // operations on that instance.
+            ExternalRequest::Stop => {
+                self.allowed.migrate_as_target =
+                    Disposition::Deny(DenyReason::HaltPending);
+                self.allowed.start = Disposition::Deny(DenyReason::HaltPending);
+                self.allowed.migrate_as_source =
+                    Disposition::Deny(DenyReason::HaltPending);
+                self.allowed.reboot =
+                    Disposition::Deny(DenyReason::HaltPending);
+                self.allowed.stop = Disposition::Ignore;
+            }
+        }
+
+        self.queue.push_back(request);
+        Ok(())
+    }
+
+    /// Notifies the queue that its instance is now running. This allows
+    /// requests to reboot and migrate out of the instance.
+    pub fn notify_instance_running(&mut self) {
+        info!(
+            self.log,
+            "Instance is running, allowing migration out and reboot"
+        );
+
+        self.allowed.migrate_as_source = RequestDisposition::Enqueue;
+        self.allowed.reboot = RequestDisposition::Enqueue;
+    }
+
+    /// Notifies the queue that its instance has stopped. This blocks all new
+    /// requests except requests to stop, which are ignored.
+    pub fn notify_instance_stopped(&mut self) {
+        info!(self.log, "Instance is stopped, denying all actions");
+        self.allowed.migrate_as_target =
+            RequestDisposition::Deny(RequestDeniedReason::InstanceNotActive);
+        self.allowed.start =
+            RequestDisposition::Deny(RequestDeniedReason::InstanceNotActive);
+        self.allowed.migrate_as_source =
+            RequestDisposition::Deny(RequestDeniedReason::InstanceNotActive);
+        self.allowed.reboot =
+            RequestDisposition::Deny(RequestDeniedReason::InstanceNotActive);
+        self.allowed.stop = RequestDisposition::Ignore;
+    }
+
+    /// Notifies the queue that its instance has failed. This blocks all new
+    /// requests except requests to stop, which are ignored.
+    pub fn notify_instance_failed(&mut self) {
+        info!(self.log, "Instance has failed, denying all actions");
+        self.allowed.migrate_as_target =
+            RequestDisposition::Deny(RequestDeniedReason::InstanceFailed);
+        self.allowed.start =
+            RequestDisposition::Deny(RequestDeniedReason::InstanceFailed);
+        self.allowed.migrate_as_source =
+            RequestDisposition::Deny(RequestDeniedReason::InstanceFailed);
+        self.allowed.reboot =
+            RequestDisposition::Deny(RequestDeniedReason::InstanceFailed);
+    }
```

Comment:
We could also do something similar here -- return a new AllowedRequests struct to wholesale replace the existing one.